source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveContainerRenderer.cs

```diff
@@ -76,69 +76,109 @@ public static FrameworkElement Render(AdaptiveContainer container, AdaptiveRende
                     if (cardElement.Separator && uiContainer.Children.Count > 0)
                     {
-                        Grid separator = AddSeparator(context, cardElement, uiContainer);
-                        tag = new TagContent(separator, uiContainer);
+                        separator = AddSeparator(context, cardElement, uiContainer);
                     }
                     else if (uiContainer.Children.Count > 0)
                     {
-                        var spacing = context.Config.GetSpacing(cardElement.Spacing);
-                        Thickness renderedMargin = uiElement.Margin;
-                        uiElement.Margin = new Thickness(renderedMargin.Left,
-                                                         renderedMargin.Top + spacing,
-                                                         renderedMargin.Right,
-                                                         renderedMargin.Bottom);
-                        tag = new TagContent(cardElement.Spacing, uiContainer);
+                        separator = AddSpacing(context, cardElement, uiContainer);
                     }
 
+                    tag = new TagContent(separator, uiContainer);
+
                     uiElement.Tag = tag;
 
+                    // Sets the minHeight property for Container and ColumnSet
                     if (cardElement.Type == "Container" || cardElement.Type == "ColumnSet")
                     {
                         AdaptiveCollectionElement collectionElement = (AdaptiveCollectionElement)cardElement;
                         uiElement.MinHeight = collectionElement.PixelMinHeight;
                     }
 
+                    int rowDefinitionIndex = uiContainer.RowDefinitions.Count;
+                    RowDefinition rowDefinition = null;
                     if (cardElement.Height != AdaptiveHeight.Stretch)
                     {
-                        context.SetVisibility(uiElement, cardElement.IsVisible, tag);
-                        uiContainer.RowDefinitions.Add(new RowDefinition() { Height = GridLength.Auto });
-                        Grid.SetRow(uiElement, uiContainer.RowDefinitions.Count - 1);
+                        rowDefinition = new RowDefinition() { Height = GridLength.Auto };
+
+                        uiContainer.RowDefinitions.Add(rowDefinition);
+                        Grid.SetRow(uiElement, rowDefinitionIndex);
                         uiContainer.Children.Add(uiElement);
+
+                        // Row definition is stored in the tag for containers and elements that stretch
+                        // so when the elements are shown, the row can have it's original definition,
+                        // while when the element is hidden, the extra space is not reserved in the layout
+                        tag.RowDefinition = rowDefinition;
+                        tag.ViewIndex = rowDefinitionIndex;
+
+                        context.SetVisibility(uiElement, cardElement.IsVisible, tag);
```

Comment:
can you refactor this with the stuff in the else statement to reduce the duplication?